arch/x86/core/userspace.S

```diff
@@ -21,10 +21,180 @@ GTEXT(z_arch_user_string_nlen_fixup)
 /* Imports */
 GDATA(_k_syscall_table)
 
+#ifdef CONFIG_X86_KPTI
+/* Switch from the shadow to the kernel page table, switch to the interrupted
+ * thread's kernel stack, and copy all context from the trampoline stack.
+ *
+ * Assumes all registers are callee-saved since this gets called from other
+ * ASM code. Assumes a particular stack layout which is correct for
+ * _exception_enter and _interrupt_enter when invoked with a call instruction:
+ *
+ *  28 SS
+ *  24 ES
+ *  20 EFLAGS
+ *  16 CS
+ *  12 EIP
+ *  8  isr_param or exc code
+ *  4  isr or exc handler
+ *  0  return address
+ */
+SECTION_FUNC(TEXT, z_x86_trampoline_to_kernel)
+	/* Check interrupted code segment to see if we came from ring 3
+	 * and hence on the trampoline stack
+	 */
+	testb $3, 16(%esp) /* Offset of CS */
+	jz 1f
+
+	/* Stash these regs as we need to use them */
+	pushl	%esi
+	pushl	%edi
+
+	/* Switch to kernel page table */
+	movl	$z_x86_kernel_pdpt, %esi
+	movl	%esi, %cr3
+
+	/* %esi = _kernel->current->stack_info.start
+	 *
+	 * This is the lowest address of the user mode stack, and higest
+	 * address of the kernel stack, they are adjacent.
+	 * We want to transplant context here.
+	 */
+	movl	$_kernel, %esi
+	movl	_kernel_offset_to_current(%esi), %esi
+	movl	_thread_offset_to_stack_start(%esi), %esi
```

Comment:
Can put that offset into the use of ESI two instructions later and eliminate this load instruction:

    movl _thread_offset_to_stack_start(%esi), %esp

Reply:
good idea!